packages/app/src/utils/dynamicUI/extractDynamicConfig.ts

```diff
@@ -255,6 +280,106 @@ function extractDynamicConfig(
+          const mi = dr.menuItem;
+          if (isStaticPath(dr.path)) {
+            itemsMap[itemName] = {
+              name: itemName,
+              icon: mi && 'icon' in mi && mi.icon ? mi.icon : '',
+              title: mi && 'text' in mi && mi.text ? mi.text : '',
+              to: dr.path ?? '',
+              priority: 0,
+              children: [],
+            };
+          }
+        });
+      }
+
+      // Process menuItems
+      if (customProperties.menuItems) {
+        Object.entries(customProperties.menuItems).forEach(
+          ([key, menuItemObj]) => {
+            if (!itemsMap[key]) {
+              itemsMap[key] = {
+                name: key,
+                title: menuItemObj.title ?? '',
+                icon: menuItemObj.icon ?? '',
+                priority: menuItemObj.priority ?? 0,
+                to: '',
+                children: [],
+              };
+            } else {
+              // Update the existing item with any new properties from the menuItem
+              itemsMap[key].title = menuItemObj.title ?? itemsMap[key].title;
+              itemsMap[key].icon = menuItemObj.icon ?? itemsMap[key].icon;
+              itemsMap[key].priority = Math.max(
+                itemsMap[key].priority ?? 0,
+                menuItemObj.priority ?? 0,
+              );
+              itemsMap[key].to = itemsMap[key].to ?? '';
+            }
+
+            if (menuItemObj.parent) {
+              // Ensure the parent exists in itemsMap
+              if (!itemsMap[menuItemObj.parent]) {
+                itemsMap[menuItemObj.parent] = {
+                  name: menuItemObj.parent,
+                  title: '',
+                  icon: '',
+                  priority: 0,
+                  children: [],
+                };
+              }
+
+              // Add the current item as a child of its parent
+              itemsMap[menuItemObj.parent].children.push(itemsMap[key]);
+
+              // Remove the child from the top level, if it exists
+              delete itemsMap[key];
+            }
+          },
+        );
+      }
+
+      // Merge itemsMap with pluginSet
+      Object.values(itemsMap).forEach(item => {
+        const existingItem = pluginSet.find(
+          existing => existing.name === item.name,
+        );
+
+        // If the item is already in pluginSet, merge children
+        if (existingItem) {
+          existingItem.children = [
+            ...(existingItem.children ?? []),
+            ...(item.children ?? []),
+          ];
+          existingItem.icon =
+            item.icon.length > 0 ? item.icon : existingItem.icon;
+          existingItem.title =
+            item.title.length > 0 ? item.title : existingItem.title;
+          existingItem.priority = Math.max(
+            existingItem.priority ?? 0,
+            item.priority ?? 0,
+          );
+        } else {
+          pluginSet.push(item);
+        }
+      });
+
+      return pluginSet;
```

Comment:
You return a set of plugins? Or a list of menu items?

Feel free to extract this code even more into its own file and please add some unit tests for it.

I would recommend to build this in two steps:

1. Extract all menu items
2. Build the menu tree

From a short look, I'm unsure if this code works also with reference across plugins:

Case 1:

```yaml
- name: plugin 1
  menuItems:
    admin:
      title: Administrator
    entry:
      title: Menu under Admin from plugin 1
      parent: admin
- name: plugin 2
  menuItems:
    entry:
      title: Menu under Admin from plugin 2
      parent: admin
```

Case 2:

```yaml
- name: plugin 1
  menuItems:
    entry:
      title: Menu under Admin from plugin 1
      parent: admin
- name: plugin 2
  menuItems:
    admin:
      title: Administrator
    entry:
      title: Menu under Admin from plugin 2
      parent: admin
```

Case 3:

Admin is defined in both plugins.

Reply:
Different plugins can share same parent menu item config, but they won't have same plugin menu item as the `key` of `menuItems` must match the unique path defined in `dynamicRoutes`. In case 1 and 2, both plugin 1 and plugin 2 use same key of `entry`, which won't happen. But the following case works.
```
- name: plugin 1
  menuItems:
    admin:
      title: Administrator
    plugin-1:
      title: Menu under Admin from plugin 1
      parent: admin
- name: plugin 2
  menuItems:
    plugin-2:
      title: Menu under Admin from plugin 2
      parent: admin
```

Case 3 also works. The later definition of parent menu item overrides the former one.